vsc/baseline/dns/layers.py

```diff
@@ -0,0 +1,257 @@
+import math
+import torch
+import numpy as np
+import torch.nn as nn
+import torch.nn.functional as F
+
+from einops import rearrange
+
+
+class L2Constrain(object):
+    def __init__(self, axis=-1, eps=1e-6):
+        self.axis = axis
+        self.eps = eps
+
+    def __call__(self, module):
+        if hasattr(module, "weight"):
+            w = module.weight.data
+            module.weight.data = F.normalize(w, p=2, dim=self.axis, eps=self.eps)
+
+
+class Attention(nn.Module):
```

Comment:
What do you think about using these models in torchscript format, so we don't need to include code for the layers? (Some pytorch code is difficult to use as a torchscript, but a lot of code works out of the box this way.)

Just an idea, but that's what we do for SSCD to avoid adding implementations here.

Reply:
For DnS, that would work and simplify the code significantly.

But for DINO, it can not be easily converted to torchscript. So some code for this baseline is necessary.

Reply:
What is the failure with DINO? I believe I've converted DINO to torchscript in the past for the SSCD paper, but I no longer remember if I had to do anything tricky to accomplish this. :)